src/debug_utils-inl.h

```diff
@@ -27,13 +27,38 @@ struct ToStringHelper {
   }
   static std::string Convert(const std::string& value) { return value; }
   static std::string Convert(bool value) { return value ? "true" : "false"; }
+  template <unsigned BASE_BITS,
+            typename T,
+            typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+  static std::string BaseConvert(T value, bool upper) {
+    std::string ret;
+    const char* digits = upper ? "0123456789ABCDEF" : "0123456789abcdef";
+    do {
+      unsigned digit = value & ((1 << BASE_BITS) - 1);
+      ret.push_back(
+          (BASE_BITS < 4 ? static_cast<char>('0' + digit) : digits[digit]));
+    } while ((value >>= BASE_BITS) != 0);
+    std::reverse(ret.begin(), ret.end());
+    return ret;
+  }
+  template <unsigned BASE_BITS,
+            typename T,
+            typename std::enable_if<!std::is_integral<T>::value, int>::type = 0>
+  static std::string BaseConvert(T value, bool upper) {
```

Comment:
Boolean trap. It's not exactly wrong but it'd be a little nicer to use an enum to make it clear at call sites what the purpose of the parameter is.

Reply:
I think it's necessary because this is inner implementation, the caller doesn't care about this. Add `enum` is more redundant

Reply:
@Himself65 I think using an enum would help with code legibility, even though a `bool` is sufficient. [Here](https://github.com/nodejs/node/blob/c16007381f0325af866e9f382b0d9457af540d08/src/js_native_api_v8.cc#L405-L413)'s an example from N-API.

OTOH do we need `upper` at all? Can't we just use `std::to_upper` at the [call site](https://github.com/nodejs/node/pull/32385/files#diff-01bc13b1f3ab5680803f6cc2f3d84884R103) with an iterator as given in https://en.cppreference.com/w/cpp/string/byte/toupper? Maybe add a helper `std::string ToUpper(std::string src);` if we don't already have something that does this somewhere in the source tree?

Reply:
We do have `ToUpper()` in `util.h`, yes :)

Reply:
@Himself65 yaaay! We have https://github.com/nodejs/node/blob/c16007381f0325af866e9f382b0d9457af540d08/src/util-inl.h#L294-L299

Reply:
nice, I will handle it later